crates/attestation/src/attestation.rs

```diff
@@ -287,51 +366,86 @@ impl Attestation {
+            &expected_measurements.rtmrs.rtmr0,
+        )?;
+        compare_hashes(
+            "rtmr1_report_data",
+            &report_data.rt_mr1,
+            &expected_measurements.rtmrs.rtmr1,
+        )?;
+        compare_hashes(
+            "rtmr2_report_data",
+            &report_data.rt_mr2,
+            &expected_measurements.rtmrs.rtmr2,
+        )?;
+        compare_hashes(
+            "mrtd_report_data",
+            &report_data.mr_td,
+            &expected_measurements.rtmrs.mrtd,
+        )?;
+
+        compare_hex_hashes(
+            "rtmr0_tcb_info",
+            &tcb_info.rtmr0,
+            &hex::encode(expected_measurements.rtmrs.rtmr0),
+        )?;
+        compare_hex_hashes(
+            "rtmr1_tcb_info",
+            &tcb_info.rtmr1,
+            &hex::encode(expected_measurements.rtmrs.rtmr1),
+        )?;
+        compare_hex_hashes(
+            "rtmr2_tcb_info",
+            &tcb_info.rtmr2,
+            &hex::encode(expected_measurements.rtmrs.rtmr2),
+        )?;
+        compare_hex_hashes(
+            "mtrd_tcb_info",
+            &tcb_info.mrtd,
+            &hex::encode(expected_measurements.rtmrs.mrtd),
+        )?;
+
+        Ok(())
     }
 
     /// Verifies RTMR3 by replaying event log.
-    fn verify_rtmr3(&self, report_data: &dcap_qvl::quote::TDReport10, tcb_info: &TcbInfo) -> bool {
-        tcb_info.rtmr3 == hex::encode(report_data.rt_mr3)
-            && Self::verify_event_log_rtmr3(&tcb_info.event_log, report_data.rt_mr3)
+    fn verify_rtmr3(
+        &self,
+        report_data: &dcap_qvl::quote::TDReport10,
+        tcb_info: &TcbInfo,
+    ) -> Result<(), VerificationError> {
+        compare_hex_hashes("rtmr3", &tcb_info.rtmr3, &hex::encode(report_data.rt_mr3))?;
+
+        Self::verify_event_log_rtmr3(&tcb_info.event_log, report_data.rt_mr3)
     }
 
     /// Verifies app compose configuration and hash. The compose-hash is measured into RTMR3, and
     /// since it's (roughly) a hash of the unmeasured docker_compose_file, this is sufficient to
     /// prove its validity.
-    fn verify_app_compose(&self, tcb_info: &TcbInfo) -> bool {
-        let app_compose: AppCompose = match serde_json::from_str(&tcb_info.app_compose) {
-            Ok(compose) => compose,
-            Err(e) => {
-                tracing::error!("Failed to parse app_compose JSON: {:?}", e);
-                return false;
-            }
-        };
+    fn verify_app_compose(&self, tcb_info: &TcbInfo) -> Result<(), VerificationError> {
+        let app_compose: AppCompose = serde_json::from_str(&tcb_info.app_compose)
+            .map_err(|e| VerificationError::AppComposeParsing(e.to_string()))?;
 
-        let mut events = tcb_info
-            .event_log
-            .iter()
-            .filter(|event| event.event == COMPOSE_HASH_EVENT && event.imr == RTMR3_INDEX);
-
-        let payload_is_correct = events.next().is_some_and(|event| {
-            event.event_payload == tcb_info.compose_hash
```

Comment:
I think this line is no longer verified in the new implementation, at least not explicitly as it was before. We compare the hash in `validate_app_compose_payload`, but `event_payload` is not used as far I can see